contracts/BSCValidatorSet.sol

```diff
@@ -468,17 +556,179 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+    if (averageDistribute != 0) {
+      for (uint i = 0; i < index; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+      uint n = currentValidatorSet.length;
+      for (uint i = index + 1; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return index;
+  }
+
+  function _felony(address validator, uint256 index) private returns (bool){
+    uint256 income = currentValidatorSet[index].incoming;
+    uint256 rest = currentValidatorSet.length - 1;
+    if (getValidators().length <= 1) {
+      // will not remove the validator if it is the only one validator.
+      currentValidatorSet[index].incoming = 0;
+      return false;
+    }
+    emit validatorFelony(validator, income);
+
+    // remove the validator from currentValidatorSet
+    delete currentValidatorSetMap[validator];
+    // It is ok that the validatorSet is not in order.
+    if (index != currentValidatorSet.length - 1) {
+      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length - 1];
+      validatorExtraSet[index] = validatorExtraSet[currentValidatorSet.length - 1];
+
+      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
+    }
+    currentValidatorSet.pop();
+    validatorExtraSet.pop();
+
+    uint256 averageDistribute = income / rest;
+    if (averageDistribute != 0) {
+      uint n = currentValidatorSet.length;
+      for (uint i = 0; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+    return true;
+  }
+
+  function _forceMaintainingValidatorsExit(Validator[] memory _validatorSet) private returns (Validator[] memory unjailedValidatorSet){
+    uint256 numOfFelony = 0;
+    address validator;
+    bool isFelony;
+
+    // 1. validators exit maintenance
+    for (uint i = 0; i < currentValidatorSet.length; i++) {
+      if (!validatorExtraSet[i].isMaintaining) {
+        continue;
+      }
+
+      // only maintaining validators
+      validator = currentValidatorSet[i].consensusAddress;
+
+      // exit maintenance
+      isFelony = _exitMaintenance(validator, i);
+      if (!isFelony || numOfFelony >= _validatorSet.length - 1) {
+        continue;
+      }
+
+      // record the jailed validator in validatorSet
+      for (uint index = 0; index < _validatorSet.length; index++) {
+        if (_validatorSet[index].consensusAddress == validator) {
+          _validatorSet[index].jailed = true;
+          numOfFelony++;
+          break;
+        }
+      }
+    }
+
+    // 2. clear all maintain info
+    numOfMaintaining = 0;
```

Comment:
Duplicated code here, since `doupdate`  will do this.

Reply:
fixed